### PR TITLE
Global game state machine

### PR DESCRIPTION
**Pull request by JunkyX1122**
Basic frame for the state machines for the client and server.
Server has 3 states:

Busy (loading shit, base state)
Waiting (loaded shit, no players)
Running (loaded shit, players)
Client has 5:

Inactive (base state)
Loading (load shit, happens between most transitions between the below states)
Menu
InGameplay (the game)
You can change the state of each by using SetServerState/SetClientState, and what state is updated is automatically set.
Also Joe told me to change the Update in the base state class so if any state machine in the future doesn't work please talk to him.
There is also a vector for multiple states which are meant to be levels but it isn't used anywhere anyways because idk how we're gonna handle that.

===NEW===
Client state machine has been updated to function in game.
Pretty much all the contents of Tutorial game have been copied over into it.
All states pertaining to the client now require a reference to a gameworld and renderer upon creation.
The creation of these two are handled within the initialisation of the client state machine.
The client state machine now displays what state the machine is in on gamescreen, with what button to press to switch states.
InGame state now has a method to specify an exit state, with ExitState::Invalid preventing the game from ending.
I've also moved the states to the client folder from the 8053CoreClasses folder.

==NOTE==
Client main now no longer makes a tutorial game.

### CODE REVIEW

File: Client/ClientGameStateMachine.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#include "GameStateInactive.h"
+#include "GameStateLoading.h"
+#include "GameStateMenu.h"
+#include "GameStateInGameplay.h"
+using namespace NCL;
+using namespace CSC8503;
+ClientGameStateMachine::ClientGameStateMachine() : StateMachine()
+{
+	gameWorld = new GameWorld();
+	renderer = new GameTechRenderer(*gameWorld);
```
Comment (St4rman) [inline Client/ClientGameStateMachine.cpp:13]: Decide if you want the renderer to be made BEFORE reaching this state or at this point

File: Client/ClientGameStateMachine.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+void ClientGameStateMachine::InitialiseClientStateMachine()
+{
+	//*
+	Inactive* clientInactive = new Inactive(renderer, gameWorld);
+	Loading* clientLoading = new Loading(renderer, gameWorld);
+	Menu* clientMenu = new Menu(renderer, gameWorld);
+	InGameplay* clientInGameplay = new InGameplay(renderer, gameWorld);
+
+	StateTransition* inactiveToLoading = new StateTransition(clientInactive, clientLoading, [&](void)->bool {
```
Comment (St4rman) [inline Client/ClientGameStateMachine.cpp:30]: Decide how to move between states;  either bools or events or enum

File: Client/ClientGameStateMachine.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			}
+			void SetClientState(ClientStates clientState)
+			{
+				currentClientState = clientState;
+			}
+		protected:
+
+
+			ClientStates currentClientState;
+			std::vector<State*> availableLevels;
```
Comment (St4rman) [inline Client/ClientGameStateMachine.h:31]: Level loaders available so can be chnaged to a level loader *

File: Client/GameStateInGamplay.cpp [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+}
+void InGameplay::OnExit() {
+	this->~InGameplay();
+	world->ClearAndErase();
+	renderer->Render();
+}
+
+int InGameplay::ExitType() {
+
+	if (Window::GetKeyboard()->KeyDown(KeyboardKeys::L))
```
Comment (St4rman) [inline Client/GameStateInGamplay.cpp:51]: this condition will be like on trigger when win

File: Client/GameStateInGamplay.cpp [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+	{
+		return ExitStates::Win;
+	}
+	return ExitStates::Invalid;
+}
+
+void InGameplay::Update(float dt) {
```
Comment (St4rman) [inline Client/GameStateInGamplay.cpp:58]: Mostly tutorial game

File: Client/Renderer.h [hunk 6]
```diff
@@ -84,16 +84,15 @@ namespace NCL {
 
 			GLuint lineVAO;
 			GLuint lineVertVBO;
-			size_t lineCount;
+			size_t lineCount = 0;
```
Comment (St4rman) [inline Client/Renderer.h:87]: ed and matthew are so smart

File: Client/main.cpp [hunk 7]
```diff
@@ -27,14 +42,19 @@ int main() {
     w->ShowOSPointer(false);
     w->LockMouseToWindow(true);
 
-    TutorialGame* g = new TutorialGame();
+    //TutorialGame* g = new TutorialGame();
```
Comment (St4rman) [inline Client/main.cpp:45]: :D

File: Replicated/ServerGameStateMachine.cpp [hunk 8]
```diff
@@ -0,0 +1,38 @@
+#include "ServerGameStateMachine.h"
+#include "State.h"
+#include "StateTransition.h"
+#include "GameStateBusy.h"
+#include "GameStateWaitingPlayers.h"
+#include "GameStateRunning.h"
+using namespace NCL;
+using namespace CSC8503;
+ServerGameStateMachine::ServerGameStateMachine() : StateMachine()
```
Comment (St4rman) [inline Replicated/ServerGameStateMachine.cpp:9]: Empty servergamestate class

File: Replicated/ServerGameStateMachine.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+#include "StateMachine.h"
+#include <vector>
+namespace NCL {
```
Comment (St4rman) [inline Replicated/ServerGameStateMachine.h:4]: empty server lass

**Review by St4rman — COMMENTED** [review]
Looks relatively good, decide renderer, and check if we want enum/bool on win conditions.
Good job, less files next commit pls :]

**Review by Odycious — APPROVED** [review]
Nice setup for the game state system.

**Pull request closed without merging**